### DISCUSSION TITLE
Frontend Cannot Locate Backend Route for Login Endpoint (404 Error), any help?

### DISCUSSION THREAD

**Discussion by salitmp** [community/community#101989] [Programming Help]
### Body

script defined in package.json
  "scripts": {
    "start": "node backend/server.js",
    "server": "nodemon backend/server.js",
    "client": "npm run dev --prefix frontend",
    "dev": "concurrently \"npm run server\" \"npm run client\""
  },
------------------------
frontend/vite.config.js

import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()],
  server: {
    port:3000,
    proxy: {
      'api': {
        target: 'http://localhost:5000',
        changeOrigin: true
      }
    }
  }
})
--------------------------

backend/server.js

app.use('/api/users', userRoutes);
-------------------------------------
login controller function at 
backend/controllers/userController.js

import asyncHandler from 'express-async-handler'
import User from '../models/userModel.js';

    const userlogin = asyncHandler(async(req, res) => {
        const { email, password } = req.body;
        const user = await User.findOne({ email });
    
        if (!user) {
            res.status(400);
            throw new Error('Email does not exist.');
        }
    
        const isMatch = await bcrypt.compare(password, user.password);
        if (!isMatch) {
            res.status(400);
            throw new Error('Password is incorrect.');
        }
    
        console.log(user);
    
        const refresh_token = createRefreshToken({ id: user._id });
        res.cookie('refreshtoken', refresh_token, {
            httpOnly: true,
            path: '/',
            maxAge: 7 * 24 * 60 * 60 * 1000, // 7 days in milliseconds
        });
    
        res.json({ msg: 'Login Success' });
    });

------------------------------

backend/routers/userRouters.js

import express from  'express';
const router = express.Router();
import { userlogin } from '../controllers/userController.js';
router.post('/login', userlogin);
export default router;

----------------------------
Redux used at frontend
frontend/slices/authSlice.js

import { createSlice } from '@reduxjs/toolkit'

const initialState = {
    userInfo: localStorage.getItem('userInfo') ? JSON.parse(localStorage.getItem('userInfo')) : null,
}

const authSlice = createSlice({
    name: 'auth',
    initialState,
    reducers:{
        setCredentials: (state, action) => {
            state.userInfo = action.payload
            localStorage.setItem('userInfo', JSON.stringify(action.payload))
        },
        logout: (state, action) => {
            state.userInfo = null
            localStorage.removeItem('userInfo')
        },
    },
})

export const  {setCredentials, logout} = authSlice.actions
export default authSlice.reducer
-------------------------------

frontend/slices/apiSlice.js

import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react'

const baseQuery = fetchBaseQuery({ baseUrl: '' })

export const apiSlice = createApi({
    baseQuery,
    tagTypes: ['User'],
    endpoints: (builder) => ({}),
})
------------------------------------

frontend/slices/usersApiSlice.js

import { apiSlice } from './apiSlice'
const USERS_URL = '/api/users'

export const usersApiSlice = apiSlice.injectEndpoints ({
    endpoints:(builder) => ({
        userlogin: builder.mutation({
            query: (data) => ({
                url: `${USERS_URL}/login`,
                method: 'POST',
                body: data,
            }),
        }),
    }),
});

export const { useUserloginMutation } = usersApiSlice 
-------------------------------------------------------

frontend/screens/LoginScreen.jsx

import { React, useState, useEffect } from 'react'
import { Link, useNavigate } from 'react-router-dom'
import { Form , Button, Row, Col } from 'react-bootstrap'
import { useDispatch, useSelector } from 'react-redux'
import { useUserloginMutation } from '../slices/usersApiSlice'
import { setCredentials } from '../slices/authSlice'

 const LoginScreen = () => {
    const [email, setEmail] = useState('')
    const [password, setPassword] = useState('')

    const navigate = useNavigate()
    const dispatch = useDispatch()

    const [login, { isLoading }] = useUserloginMutation()

    const { userInfo } = useSelector((state) => state.auth)

    useEffect(() => {
        if (userInfo) {
            navigate('/') 
        }
    }, [navigate, userInfo])
  
    const submitHandler = async(e) =>{
        e.preventDefault()
       try {
            const res = await login ({ email, password }).unwrap()
            dispatch(setCredentials({ ...res }))
            navigate('/')
       } catch (err) {
            console.log(err?.data?.message || err.error)          
       }
    }
-----------------------------------------
console result, 

not available
﻿
VM901:6 
 POST http://localhost:3000/api/users/login 404 (Not Found)
await in (anonymous) (async)		
dispatch	@	VM901:6
submitHandler	@	LoginScreen.jsx:29
Show 29 more frames
LoginScreen.jsx:33 undefined

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by shanakasp** [comment]
try this , I have made some changers import React, { useState, useEffect } from 'react';
import { Form, Button, Row, Col } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { useUserloginMutation } from '../slices/usersApiSlice';
import { setCredentials } from '../slices/authSlice';
import { useNavigate } from 'react-router-dom';

const LoginScreen = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  const dispatch = useDispatch();
  const navigate = useNavigate();

  const { userInfo } = useSelector((state) => state.auth);

  const [login, { isLoading, isError, error }] = useUserloginMutation();

  useEffect(() => {
    if (userInfo) {
      navigate('/');
    }
  }, [navigate, userInfo]);

  const submitHandler = async (e) => {
    e.preventDefault();
    try {
      const { data } = await login({ email, password }).unwrap();
      dispatch(setCredentials(data)); // Assuming the response contains user info
      navigate('/');
    } catch (err) {
      console.error('Login error:', err);
    }
  };

  return (
    <Row className="justify-content-md-center">
      <Col xs={12} md={6}>
        <h1>Sign In</h1>
        <Form onSubmit={submitHandler}>
          <Form.Group controlId="email">
            <Form.Label>Email Address</Form.Label>
            <Form.Control
              type="email"
              placeholder="Enter email"
              value={email}
              onChange={(e) => setEmail(e.target.value)}
            ></Form.Control>
          </Form.Group>

          <Form.Group controlId="password">
            <Form.Label>Password</Form.Label>
            <Form.Control
              type="password"
              placeholder="Enter password"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
            ></Form.Control>
          </Form.Group>

          <Button type="submit" variant="primary" disabled={isLoading}>
            Sign In
          </Button>
        </Form>
      </Col>
    </Row>
  );
};

export default LoginScreen;
